### ISSUE TITLE
[Proposal] Tutorials

### ISSUE THREAD

**Issue by pseudo-rnd-thoughts** [Farama-Foundation/Gymnasium#28]
### Proposal

To encourage the use of Gymnasium and build up the RL community, I would propose that a large range of tutorials are created. 

This is a list of tutorials that could be made

- [ ] Implementation of a custom environment
- [ ] DQN for atari implementation (doesn't need to be fast)
- [ ] Training an agent for the new Phys3d / Phys2d environments (brax) using pytorch or jax
- [ ] Frozenlake training results with different map sizes
- [ ] Gym vectorisation (gym.vector.make)
- [ ] Training agents for the blackjack environment, it has Tuple observation space
- [ ] How to use the action sample masking, with example from Taxi
- [ ] Car racing, comparing agents with continuous and discrete action spaces
- [ ] Training a deep RL agent with pytorch (with the new step API)
- [ ] Using gym make, spec, register and registry
- [ ] Understand the gym spaces
- [ ] Exploring the impact of bipedal walker `hardcore` parameter on agent performance
- [ ] Train an agent with stable-baselines-3
- [ ] Train an agent with tianshou 
- [ ] Using gym for Multi-agent RL
- [ ] Using gym for Offline RL with D4RL
- [ ] How to render an environment
- [ ] How to develop an environment using the gym MDP API
- [ ] Using the gym wrappers and developing custom wrappers
- [ ] visualising your agent with gym.utils.play
- [ ] Understanding the gym environment checker
- [ ] Experimenting with classic control reset options random state bounds

**Comment by kir0ul** [comment]
Will there be a required format for the tutorials?

**Comment by kir0ul** [comment]
Isn't the first tutorial on _Implementation of a custom environment_ already done in: https://gymnasium.farama.org/content/environment_creation/ ?

**Comment by pseudo-rnd-thoughts** [comment]
We are using the [sphinx-gallery](https://sphinx-gallery.github.io/stable/index.html) project, therefore, we expect single python files with figures stored in a folder with the same name. 
We are in the process of transferring our current tutorials to this style but [pytorch](https://github.com/pytorch/tutorials) use this style for their tutorials

**Comment by BalajiAI** [comment]
Since I'm a beginner, I'll take DQN for Atari tutorial.

**Comment by kir0ul** [comment]
I'm working on the Frozenlake one, still work in progress but I hope to have it ready soonish.

**Comment by siddarth-c** [comment]
I am working on "Training a deep RL agent with pytorch (with the new step API)".

**Comment by hahas94** [comment]
Could someone make an update on this proposal?
For instance, which tasks are done, which ones are work in progress, so that one may know what to work on.

**Comment by pseudo-rnd-thoughts** [comment]
> Could someone make an update on this proposal? For instance, which tasks are done, which ones are work in progress, so that one may know what to work on.

Thanks, have updated

**Comment by kir0ul** [comment]
@pseudo-rnd-thoughts also the _Frozenlake training results with different map sizes_ tutorial is done.

**Comment by hahas94** [comment]
Thanks!

I am thinking to work on `Training a deep RL agent with pytorch fromscratch`. However, doesn't the REINFORCE implementation fit that requirement?

**Comment by pseudo-rnd-thoughts** [comment]
Good point @hahas94 I think we are always happy for more tutorials. You can work on another tutorial if you want

**Comment by hahas94** [comment]
Okay, if similar tutorials is a good thing I will stick to my aforementioned choice. 
I am thinking of writing a tutorial on implementing the C51 agent in PyTorch.

**Comment by jonnytracker** [comment]
We need this. For practical use case. Need external environment tutorials not another memory environment

**Comment by dantp-ai** [comment]
I can add a tutorial for "Train an agent with tianshou". The choice of agent and env is free right, as long as env is from Gymnasium?

**Comment by Kallinteris-Andreas** [comment]
@dantp-ai we have chosen to link tutorials, who purpose is to show how to use external libraries, to 

https://gymnasium.farama.org/main/tutorials/third-party-tutorials/#third-party-tutorials-about-libraries-that-use-gymnasium

**Comment by dantp-ai** [comment]
@Kallinteris-Andreas I see, so for "Train an agent with tianshou" it would simply mean listing a link to the library with a brief explanation of its focus (similar to AgileRL in the section).

Regarding, the tutorial "How to use the action sample masking, with example from Taxi", I can tackle this one. Do I understand correctly that for this tutorial we want to show the impact of the RL agent learning, when we use at each state the `info["action_mask"]` to pick up an action that results in a state change versus when we are not using the `action_mask` info?

**Comment by Kallinteris-Andreas** [comment]
1. It does mean that now
2. @pseudo-rnd-thoughts, can elaborate

**Comment by pseudo-rnd-thoughts** [comment]
@dantp-ai yes that is general idea. 
Both a discussion of what action masking is, how it is impacted in gymnasium with space.sample(mask) and a small training implementation

**Comment by dantp-ai** [comment]
@pseudo-rnd-thoughts Added the tutorial "How to use the action sample masking, with example from Taxi". Here's the PR: https://github.com/Farama-Foundation/Gymnasium/pull/1345
